File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    List<List<String>>l=new ArrayList();
    public List<List<String>> partition(String s) {
        backtrack(new ArrayList<>(),0,s);
        return l;
    }
    public void backtrack(ArrayList<String>row,int i,String s){
        if(i==s.length()){
            l.add(new ArrayList<>(row));
            return;
        }
        for(int j=i;j<s.length();j++){
            if(isPalindrome(s,i,j)){
                row.add(s.substring(i,j+1));
                backtrack(row,j+1,s);
                row.remove(row.size()-1);
            }
        }
    }
    public boolean isPalindrome(String s,int i,int j){
        while(i<=j){
            if(s.charAt(i++)!=s.charAt(j--)){
                return false;
            }
        }
        return true;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    
    bool is_palindrome(int i,int j,string &s)
    {
        while(i<j)
        {
            if(s[i++]!=s[j--]) return false;
        }
        return true;
    }
    void solve(int i,string& s,vector<string>& v,vector<vector<string>>& ans,vector<vector<int>>& dp)
    {
        if(i>=s.size())
        {
            ans.push_back(v); return;
        }
        for(int j=i;j<s.size();j++)
        {
            if(dp[i][j])
            {
                v.push_back(s.substr(i,j-i+1));
                solve(j+1,s,v,ans,dp);
                v.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        std::ios_base::sync_with_stdio(false);
        std::cout.tie(nullptr);
        std::cin.tie(nullptr);
        vector<vector<int>> dp(s.size(),vector<int>(s.size(),0));
        for(int i=0;i<s.size();i++)
        {
            for(int j=i;j<s.size();j++)
            {
                if(is_palindrome(i,j,s)) dp[i][j]=1;
            }
        }
        vector<vector<string>> ans;
        if(s.empty()) return ans;
        vector<string> v;
        solve(0,s,v,ans,dp);
        return ans;
    }
};
